{% extends "report_base.html" %} {% block body %}
<div class="logo-container">
  <img src="{{ url_for('static', filename='icons/antibody-multi-96.png') }}" alt="Antibody group">
  <img style="transform: scale(0.4); transform: translateY(50%); position: absolute" src="{{ url_for('static', filename='images/dna.png') }}" alt="DNA">
  <img style="transform: scale(1%); opacity: 0.15; transform: translateY(70%); transform: translateX(10%); position: absolute" src="{{ url_for('static', filename='images/group_antibodies.png') }}" alt="Antibody group 2">
  <span class="logo-container">IGHV mutationsstatus och subsetanalys i Kronisk lymfatisk leukemi (KLL) -<br><br><i>{{sample_id}}</i><br><br></span>
  <table class="logo-container">
    <tr>
      <td class="top_report_key">Patientnamn</td>
      <td class="top_report_val">{{ clarity_data.get('Patient Name', '') }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Personnummer</td>
      <td class="top_report_val">{{ clarity_data.get('Personal Identity Number', '') }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Prov-ID</td>
      <td class="top_report_val">{{ sample_id }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Registreringsdatum</td>
      <td class="top_report_val">{{ clarity_data.get('Date of arrival', '') }}</td>
    </tr> 
    <tr>
      <td class="top_report_key">Provtyp</td>
      <td class="top_report_val">Bone marrow / DNA</td>
    </tr><tr>
      <td class="top_report_key">Frågeställning</td>
      <!--<td class="top_report_val">{{ clarity_data.get('Diagnosis', '') }}</td>-->
      <td class="top_report_val">IGHV mutationsstatus och subsetanalys i KLL</td>
    </tr>
    <tr>
      <td class="top_report_key">Rapportdatum</td>
      <td class="top_report_val">{{ report_date }}</td>
    </tr>  
    <tr>
      <td class="top_report_key">Analysmetod</td>
      <!--<td class="top_report_val">{{ clarity_data.get('Analysis', '') }}</td>-->
      <td class="top_report_val">LymphoTrack Dx IGHV Leader Somatic Hypermutation Assay (Invivoscribe) med efterföljande analys i IMGT/VQUEST</td>
    </tr>
    <tr>
      <td class="top_report_key">Analys genomförd av</td>
      <td class="top_report_val">{{ config['PDF_ANALYSIS_RUN_AT'] }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Rapport skapad av</td>
      <td class="top_report_val">{{ current_user.get_fullname() }}</td>
    </tr>  
    <tr>
      <td class="top_report_key">Rapport-ID</td>
      <td class="top_report_val">{{ report_id }}</td>
    </tr>
    <tr>
      <td class="top_report_key">Rapportversion</td>
      <td class="top_report_val">{{ config['APP_VERSION'] }}</td>
    </tr>
  </table>
</div>
<div style="page-break-after: always;"></div>
<span class="report_header">Analysresultat och Slutsats</span>
<span class="report_text_medium" style="text-align: justify">
{{ report_summary|replace('\n', '<br>')|safe }}<br><br>
</span>
{% if report_with_vquest %}
  <div style="page-break-after: always;"></div>
  <div class="report_div" style="page-break-inside: avoid;">
  <span class="report_header">Detaljerade Analysresultat</span>

    {% set seq_count = 1 %}
    {% for seq_id, seq_value in results_summary.items() %} 
    {% set v_indel_identity_per = seq_value['V-REGION identity % (with ins/del events)'] %} 
    {% set v_indel_identity_nt = seq_value['V-REGION identity nt (with ins/del events)'] %}
    {% set seq_name = seq_id.split('_')[0] %}
    <p>
      <b>Sammanfattning för sekvens-ID{{ ("&nbsp;" * 7|int)|safe }}:{{ ("&nbsp;" * 4|int)|safe }}</b><i>{{ seq_name }}</i>
    </p>
    <p>
      <b>Analyserad sekvenslängd{{ ("&nbsp;" * 16|int)|safe }}&nbsp;:{{ ("&nbsp;" * 4|int)|safe }}</b
      ><i>{{ seq_value['Analysed sequence length'] }}</i>
    </p>
    <p>
      <b>Sekvensanalyskategori{{ ("&nbsp;" * 21|int)|safe }}:{{ ("&nbsp;" * 4|int)|safe }}</b
      ><i>{{ seq_value['Sequence analysis category'] }}</i>
    </p>
    <table class="report_variant">
      <tr>
        <th class="report_variant_header" colspan="4">
          <center>
            <span class="report_variant_header">{{ seq_name }} Summary</span>
          </center>
        </th>
      </tr>
      <tr>
        <td class="report_key" colspan="2" >
          Sequence ID
        </td>
        <td class="var_report_val" colspan="2">
          {{ seq_name }}
        </td>
      </tr>

      {% if v_indel_identity_per is not none %}
      <tr>
        <td class="report_key">Indel Summary</td>
        <td class="var_report_val" colspan="3" style="background-color: #FFFFE0">
          {{ seq_value['Indels if Any'] }}.<br />Indels detected are given below,<br />
          {% if seq_value['V-REGION insertions'] is not none %}
          <i>{{ seq_value['V-REGION insertions'] }}</i>
          {% endif %} {% if seq_value['V-REGION deletions'] is not none %}
          <br />
          <i>{{ seq_value['V-REGION deletions'] }}</i>
          {% endif %}
        </td>
      </tr>
      <tr>
        <td class="report_key">V-Domain functionality after removal of Indels</td>
        <td class="var_report_val" colspan="3" style="text-align: center; background-color: #FFFFE0" >
          {{ seq_value['V-DOMAIN Functionality'] }}<br />
        </td>
      </tr>
      <tr>
        <td class="report_key">Comment</td>
        <td class="var_report_val" colspan="3" style="background-color: #FFFFE0" >
          {{ seq_value['V-DOMAIN Functionality comment'] }}
        </td>
      </tr>
        {% elif v_indel_identity_per is none %}
        <tr><td class="report_key">V-Domain functionality</td>
        <td class="var_report_val" colspan="3" style="background-color: #FFFFE0">
          {{ seq_value['V-DOMAIN Functionality'] }}
        </td></tr>
        {% if 'see comment' in seq_value['V-DOMAIN Functionality'] %}
          <tr>
            <td><b>Comment</b></td>
            <td colspan="3" style="background-color: #FFFFE0">
              {{ seq_value['V-DOMAIN Functionality comment'] }}
            </td>
          </tr>
        {% endif %}
        {% if seq_value['Indel Messages'] is not none %}
      <tr>
        <td class="report_key">Messages</td>
        <td class="var_report_val" colspan="3" style="background-color: #FFFFE0">
          {{ seq_value['Indel Messages'] }}
        </td></tr>
        {% endif %}
      
        {% endif %}
      </tr>
      <tr>
        <td class="report_key">V-GENE and allele</td>
        <td class="var_report_val">{{ seq_value['V-GENE and allele'] }}</td>
        <td class="var_report_val"><b>Score:</b> {{ seq_value['V-REGION score'] }}</td>
        {% if v_indel_identity_per is not none %}
        <td class="var_report_val" style="background-color: #FFFFE0">
          <b>Identity:</b> {{ seq_value['V-REGION identity %'] }}% ({{ seq_value['V-REGION identity nt'] }}) [{{
          v_indel_identity_per }}% ({{ v_indel_identity_nt }})]
        </td>
        {% elif v_indel_identity_per is none %}
        <td class="var_report_val" style="background-color: #FFFFE0">
          <b>Identity:</b> {{ seq_value['V-REGION identity %'] }}% ({{ seq_value['V-REGION identity nt'] }})
        </td>
        {% endif %}
      </tr>
      <tr>
        <td class="report_key">J GENE and allele</td>
        <td class="var_report_val">{{ seq_value['J-GENE and allele'] }}</td>
        <td class="var_report_val"><b>Score:</b> {{ seq_value['J-REGION score'] }}</td>
        <td class="var_report_val">
          <b>Identity:</b> {{ seq_value['J-REGION identity %'] }}% ({{ seq_value['J-REGION identity nt'] }})
        </td>
      </tr>
      <tr>
        <td class="report_key">D-GEN and allele by IMGT/JunctionAnalysis</td>
        <td class="var_report_val">{{ seq_value['D-GENE and allele'] }}</td>
        <td colspan="2" class="var_report_val">
          <b>D-REGION</b> is in reading frame {{ seq_value['D-REGION reading frame'] }}
        </td>
      </tr>
      <tr>
        <td class="report_key">FR-IMGT lengths, CDR-IMGT lengths and AA JUNCTION</td>
        <td class="var_report_val">{{ seq_value['FR-IMGT lengths'] }}</td>
        <td class="var_report_val">[{{ seq_value['CDR-IMGT lengths'] }}]</td>
        <td class="var_report_val">{{ seq_value['AA JUNCTION'] }}</td>
      </tr>
      <tr>
        <td class="report_key">JUNCTION length (in nt) and decryption</td>
        <td class="var_report_val">{{ seq_value['JUNCTION-nt nb'] }} nt = {{ seq_value['JUNCTION decryption'] }}</td>
        <td colspan="2" class="var_report_val">
          <a href="https://www.imgt.org/IMGT_jcta/decryption" target="_blank ">(3'V)3'{N1}5'(D)3'{N2}5'(5'J)</a>
        </td>
      </tr>
    </table>
    <span class="report_header_medium">
      Subsetanalys
    </span>
    <span class="report_text_medium" style="text-align: justify">
      {{ seq_value['CLL Subset Summary'] }}
    </span><br><br>
    <hr>
    
    {% set seq_count = seq_count + 1 %} {% endfor %}
  </div>
{% endif %}
<span class="report_header">Analysbeskrivning</span>
<span class="report_text_medium" style="text-align: justify">

  DNA har extraherats från insänt prov (vanligtvis färskt benmärgsprov eller blodprov) och analyserats med massiv parallell sekvensering (MPS, även kallat NGS). Sekvensanalysen omfattar IGH-genen som inkluderas i LymphoTrack Dx IGHV Leader Somatic Hypermutation Assay (Invivoscribe). Efterföljande sekvensanalys genomförs då sekvenseringen uppfyller QC-kraven Analysen avser detektion av funktionellt IGH-genrearrangemang (V-, D- och J-gen), förekomst av mutationsstatus samt subsettillhörighet och utförs via webbtjänsten <a href="https://www.imgt.org/IMGT_vquest/vquest" target="_blank">IMGT/V-QUEST</a>. I de fall där den analyserade sekvensen avviker från germline-sekvensen i IMGT/V-QUEST (<97% identitet) bedöms provet vara hypermuterat (M-CLL). Om den analyserade sekvensen däremot har ≥98 identitet mot germline-sekvensen bedöms provet som icke-muterat (U-CLL). Borderlinetillhörighet anges då sekvensidentiteten faller inom 97-97.99%.  

  Detektion av subsettillhörighet görs genom analys av IGH-genes struktur mot Vidare analyseras IGH-genes struktur då denna kan påvisa subsettillhörighet och där specifika subsets delar biologiska och kliniska egenskaper. Subsetanalysen görs via IMGT/VQUEST (subset 2 och 8) och Arrest (subset 1 och 4).
</span>
<span class="report_header">References</span>
<span class="report_text_medium">
[1] Agathangelidis, A., Chatzidimitriou, A., Chatzikonstantinou, T., Tresoldi, C., Davis, Z., Giudicelli, V., ... & ERIC, the European Research Initiative on CLL. (2022). Immunoglobulin gene sequence analysis in chronic lymphocytic leukemia: the 2022 update of the recommendations by ERIC, the European Research Initiative on CLL. Leukemia, 36(8), 1961-1968.<br><br>
[2] Brochet, X., Lefranc, M. P., & Giudicelli, V. (2008). IMGT/V-QUEST: the highly customized and integrated system for IG and TR standardized VJ and VDJ sequence analysis. Nucleic acids research, 36(suppl_2), W503-W508.
</span>
{% endblock %}
